### ISSUE TITLE
Proposal: Field map syntax v2

### ISSUE THREAD

**Issue by EdSchouten** [elm/compiler#1926]
This is a continuation of #984. This issue has been closed for years, even though I think it describes a problem that is quite noticeable.

Right now the language already offers a shorthand for obtaining a getter function for a record field: `(\v -> v.foo)`. Namely, it is possible to simply write `.foo`. No such syntax is provided for generating setters or mappers. This is unfortunate, as the existing syntax (`{a | b = c}` for setting and `{a | b = f a.b }` for mapping) cannot be chained easily.

In #984, it was mentioned discussions about this had to wait until after 0.16, which is why it was closed. As 0.16 is long behind us, I think now's the right moment to give this a second chance. Instead of having a single proposal for setters and mappers, let's focus just on mappers. I think the absence of those is the most annoying, as it means nested models cannot be updated easily.

One of the open issues is/was what the syntax of such operators would need to be. In #984, various flavours were proposed. Concerns were that some of the proposals were cryptic; not obvious to novice/new users what the meaning is. My question therefore is, what about the following syntax?

```
{| fieldName }
```

Basically, just a subset of the existing field assignment syntax, with the source expression and value assignment omitted. For example, to increment `someInstance.someField.someSubField.someCounter`, you could use:

```
someInstance |> ({| someField } <| {| someSubField } <| {| someCounter } <| (+) 1)
```

Models containing `Maybe` and `Result` fields may be updated as well:

```
someInstance
    |> ({| someField } <|
             Result.map <|
                 {| someOtherField } <|
                     Maybe.map <|
                         (+) 1
       )

**Comment by evancz** [comment]
I added this idea to https://github.com/elm/compiler/issues/1375

For the purpose of decision making, I think It has been discussed quite extensively and thoroughly already in various online and offline forums. The role of "deeply nested updates" was what the whole community was talking about for maybe a year or so, so I think the folks on discourse.elm-lang.org or the Elm slack can give more background on that sort of stuff.
